internal/controllers/topology/cluster/structuredmerge/dryrun.go

```diff
@@ -0,0 +1,301 @@
+						path:     ctx.path,
+						fieldsV1: fieldV1Map,
+						modified: modifiedItem,
+						original: originalItem,
+					})
+					hasChanges = hasChanges || itemHasChanges
+					hasSpecChanges = hasSpecChanges || itemHasSpecChanges
+
+					// If there are already changes detected, it is possible to skip processing other items.
+					if hasChanges {
+						break
+					}
+				}
+				return
+			}
+
+			if isListSet(ctx.fieldsV1) {
+				s := sets.NewString()
+				for v := range ctx.fieldsV1 {
+					s.Insert(strings.TrimPrefix(v, "v:"))
+				}
+
+				for _, v := range modifiedList {
+					// NOTE: ignoring this error because API server ensures the keys in listMap are scalars value.
+					vString, _ := v.(string)
+					if !s.Has(vString) {
+						return pathToResult(ctx.path)
+					}
+				}
+				return
+			}
+		}
+
+		// NOTE: listType = atomic is managed like scalar values down in the func;
+		// a list is atomic when the corresponding FieldV1 doesn't have info for the list items.
+	}
+
+	// Otherwise, the func is processing scalar or atomic values.
+
+	// Check if the field has been added (it wasn't managed before).
+	// NOTE: This prevents false positive when handling metadata, because it is required to have metadata.name and metadata.namespace
+	// in modified, but they are not tracked as managed field.
+	notManagedBefore := ctx.fieldsV1 == nil
+	if len(ctx.path) == 1 && ctx.path[0] == "metadata" {
+		notManagedBefore = false
+	}
+
+	// Check if the field value is changed.
+	// NOTE: it is required to use reflect.DeepEqual because in case of atomic map or lists the value is not a scalar value.
+	valueChanged := !reflect.DeepEqual(ctx.modified, ctx.original)
+
+	if notManagedBefore || valueChanged {
+		return pathToResult(ctx.path)
+	}
+	return false, false
+}
+
+type dryRunInput struct {
+	// the path of the field being processed.
+	path contract.Path
+	// fieldsV1 for the current path.
+	fieldsV1 map[string]interface{}
+
+	// the original and the modified value for the current path.
+	modified interface{}
+	original interface{}
+}
+
+// pathToResult determine if a change in a path impact the spec.
+// We assume there is always a change when this call is called; additionally
+// we determine the change impacts spec when the path is the root of the object
+// or the path starts with spec.
+func pathToResult(p contract.Path) (hasChanges, hasSpecChanges bool) {
+	return true, len(p) == 0 || (len(p) > 0 && p[0] == "spec")
+}
+
+// getFieldV1Keys returns the keys for a listMap item in metadata.managedFields;
+// e.g. given the `"k:{\"field1\":\"id1\"}": {...}` item in a ListMap it returns {field1:id1}.
+func getFieldV1Keys(v string) map[string]string {
+	keys := map[string]string{}
+	keysJSON := strings.TrimPrefix(v, "k:")
+	// NOTE: ignoring this error because API server ensures this is a valid yaml.
+	_ = json.Unmarshal([]byte(keysJSON), &keys)
+	return keys
+}
+
+// getItemKeys returns the keys value pairs for an item in the list.
```

Comment:
Can we please add similar examples like we have for `getFieldV1Keys` to this func and the ones below? Thank you!

Reply:
done